ee/clickhouse/queries/cohort_query.py

```diff
@@ -474,52 +482,107 @@ def get_performed_event_first_time(self, prop: Property, prepend: str, idx: int)
+        _outer_fields.append("person_id")
+        _outer_fields.extend(self._fields)
+
+        outer_query = f"""
+        SELECT {", ".join(_outer_fields)} FROM ({intermediate_query})
+        GROUP BY person_id
+        """
+
+        return outer_query, {"team_id": self._team_id, "events": self._events, **params}, self.FUNNEL_QUERY_ALIAS
+
+    def _get_sequence_filter(self, prop: Property, idx: int) -> Tuple[List[str], List[str], List[str], Dict[str, Any]]:
+        event = validate_entity((prop.event_type, prop.key))
+        entity_query, entity_params = self._get_entity(event, "event_sequence", idx)
+        seq_event = validate_entity((prop.seq_event_type, prop.seq_event))
+
+        seq_entity_query, seq_entity_params = self._get_entity(seq_event, "seq_event_sequence", idx)
+
+        time_value = parse_and_validate_positive_integer(prop.time_value, "time_value")
+        time_interval = validate_interval(prop.time_interval)
+        seq_date_value = parse_and_validate_positive_integer(prop.seq_time_value, "time_value")
+        seq_date_interval = validate_interval(prop.seq_time_interval)
+        self._check_earliest_date((time_value, time_interval))
+
+        event_prepend = f"event_{idx}"
+
+        duplicate_event = 0
+        if event == seq_event:
+            duplicate_event = 1
+
+        aggregate_cols = []
+        aggregate_condition = f"max(if({event_prepend}_latest_0 < {event_prepend}_latest_1 AND {event_prepend}_latest_1 <= {event_prepend}_latest_0 + INTERVAL {seq_date_value} {seq_date_interval}, 2, 1)) AS {self.SEQUENCE_FIELD_ALIAS}_{self.sequence_filters_lookup[str(prop.to_dict())]}"
+        aggregate_cols.append(aggregate_condition)
+
+        condition_cols = []
+        timestamp_condition = f"min({event_prepend}_latest_1) over (PARTITION by person_id ORDER BY timestamp DESC ROWS BETWEEN UNBOUNDED PRECEDING AND {duplicate_event} PRECEDING) {event_prepend}_latest_1"
+        condition_cols.append(f"{event_prepend}_latest_0")
+        condition_cols.append(timestamp_condition)
+
+        step_cols = []
+        step_cols.append(
+            f"if({entity_query} AND timestamp > now() - INTERVAL {time_value} {time_interval}, 1, 0) AS {event_prepend}_step_0"
+        )
+        step_cols.append(f"if({event_prepend}_step_0 = 1, timestamp, null) AS {event_prepend}_latest_0")
+
+        step_cols.append(
+            f"if({seq_entity_query} AND timestamp > now() - INTERVAL {time_value} {time_interval}, 1, 0) AS {event_prepend}_step_1"
+        )
+        step_cols.append(f"if({event_prepend}_step_1 = 1, timestamp, null) AS {event_prepend}_latest_1")
+
+        return step_cols, condition_cols, aggregate_cols, {**entity_params, **seq_entity_params}
 
-        return query, params, self.FUNNEL_QUERY_ALIAS
+    def get_performed_event_sequence(self, prop: Property, prepend: str, idx: int) -> Tuple[str, Dict[str, Any]]:
+        return f"{self.SEQUENCE_FIELD_ALIAS}_{self.sequence_filters_lookup[str(prop.to_dict())]} = 2", {}
```

Comment:
Super nit: it'd be nice to push the logic check for ` = 2` into the select part of the clause - just to keep the logic consistently up there.

Reply:
oh yes